### ISSUE TITLE
purely lua module or spoon?

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/Spoons#27]
@cmsj, do we have a clear distinction for when something that is purely lua should be in a module versus a spoon?

In this specific case, I recently acquired a Phillips Hue bridge and have been playing around with it... I suspected I might need to code the local network discovery portion in Objective-C, so I started it as a module, but it turns out it can be done completely with `hs.socket.udp`, so I'm now looking at what will likely be an entirely lua based module.

A very early version is at https://github.com/asmagill/hammerspoon_asm/tree/master/phillipshue if you or anyone else wants to take a look -- no documentation for it yet, but if you're familiar with the Hue and its Rest API, it's not hard to follow.

So I'm trying to decide if it fits better as a module like `hs.milight` or as a spoon.  Thoughts?  Preferences?

**Comment by zzamboni** [comment]
My (very vague) take on this is that Spoons are mainly meant for out-of-the-box end-user functionality, since they can be easily installed and used, whereas modules are for building blocks that you are expected to use in your Lua code to achieve what you want.

However, spoons would also be a good way of distributing libraries that have little chance of making it into the main Hammerspoon core, just to make them more accessible - particularly if a spoon-installation mechanism such as https://github.com/Hammerspoon/Spoons/pull/20 ends up gaining traction.

In your case, I think `hs.hue` might be better off as a module, whereas something that uses it to, for example, set the lighting according to the time of day, with some easily-configurable options, would be a Spoon.

**Comment by cmsj** [comment]
@asmagill definitely a module I think :)

(also I have some Hue related stuff in https://github.com/cmsj/hammerspoon-config/blob/master/hueMotionSensor.lua which does the simplest form of bridge discovery, and then just enumerates/polls the Hue Motion Sensor)

**Comment by cmsj** [comment]
(to expand my thoughts - I think that modules should be things that provide fundamental units of functionality, and spoons should build on one or more modules to provide convenient user-facing utilities)
